Extract FixPath logic from Value

fsteeg opened the pull request:
As discussed in our meeting, this is an interim state for #102.

blackwinter commented on metafix/src/main/java/org/metafacture/metafix/FixMethod.java:
```diff
@@ -355,7 +355,7 @@ public void apply(final Metafix metafix, final Record record, final List<String>
         @Override
         public void apply(final Metafix metafix, final Record record, final List<String> params, final Map<String, String> options) {
-            record.transformFields(params, String::toLowerCase);
+            new FixPath(params.get(0)).transformIn(record, s -> s.toLowerCase());
```
Makes me a little sad :cry: 

FTR:

```
error: reference to transformIn is ambiguous

both method transformIn(org.metafacture.metafix.Value.Hash,java.util.function.BiConsumer<org.metafacture.metafix.Value.TypeMatcher,java.util.function.Consumer<org.metafacture.metafix.Value>>) in org.metafacture.metafix.FixPath and method transformIn(org.metafacture.metafix.Value.Hash,java.util.function.UnaryOperator<java.lang.String>) in org.metafacture.metafix.FixPath match
```

blackwinter commented on metafix/src/main/java/org/metafacture/metafix/FixPath.java:
```diff
@@ -0,0 +1,381 @@
+
+package org.metafacture.metafix;
+
+import org.metafacture.metafix.Value.Array;
+import org.metafacture.metafix.Value.Hash;
+import org.metafacture.metafix.Value.ReservedField;
+import org.metafacture.metafix.Value.TypeMatcher;
```
Hm, I'm undecided if I like these `import`s of nested types; in particular as they refer to a sibling class. Anyway, it's easier for now not to change all references.

blackwinter commented on metafix/src/main/java/org/metafacture/metafix/FixPath.java:
```diff
@@ -0,0 +1,381 @@
+            findNested(hash, field, tail(path));
+    }
+
+    /*package-private*/ Value findIn(final Value.Array array) {
```
Except for this one. Should be changed to `Array` then.

blackwinter commented on metafix/src/main/java/org/metafacture/metafix/FixPath.java:
```diff
@@ -0,0 +1,381 @@
+        array.removeIf(v -> Value.isNull(v));
+    }
+
+    protected enum InsertMode {
```
Should be `private`?

blackwinter commented on metafix/src/main/java/org/metafacture/metafix/Value.java:
```diff
@@ -256,11 +255,8 @@ public String toString() {
-            .ifArray(a -> a.transformFields(fields, operator))
-            .ifHash(h -> h.transformFields(fields, operator))
-            .orElseThrow();
+    static String[] split(final String fieldPath) {
```
`/*package-private*/`

blackwinter commented on metafix/src/main/java/org/metafacture/metafix/Value.java:
```diff
@@ -274,7 +270,7 @@ private void transformFields(final String[] fields, final UnaryOperator<String>
         private final Value value;
 
-        private TypeMatcher(final Value value) {
+        /*package-private*/ TypeMatcher(final Value value) {
```
Should be `private`.

blackwinter replied:
Not if you switch to `Value.matchType()`. See suggestions below.

fsteeg replied:
Right, saw it a moment after I commented (and deleted my comment).

blackwinter commented on metafix/src/main/java/org/metafacture/metafix/Value.java:
```diff
@@ -705,45 +533,14 @@ public Value append(final String fieldPath, final String newValue) {
         public Value get(final String field) {
             // TODO: special treatment (only) for exact matches?
-            final List<Value> list = findFields(field).map(map::get).collect(Collectors.toList());
+            final List<Value> list = matchFields(field, Stream::filter).map(map::get).collect(Collectors.toList());
```
Why did you drop the `findFields()` helper?

fsteeg replied:
I mostly tried to trim down the similar named methods, so I dropped one-liners. Feel free to add anything back that you feel is an improvement.

blackwinter commented on metafix/src/main/java/org/metafacture/metafix/FixPath.java:
```diff
@@ -0,0 +1,381 @@
+                    }
+                }
+                else {
+                    new TypeMatcher(value)
```
```suggestion
                    value.matchType()
```

blackwinter commented on metafix/src/main/java/org/metafacture/metafix/FixPath.java:
```diff
@@ -0,0 +1,381 @@
+        if (value != null) {
+            value.matchType()
+                .ifString(s -> array.set(index, operator != null ? new Value(operator.apply(s)) : null))
+                .orElse(v -> new Value.TypeMatcher(v)
```
```suggestion
                .orElse(v -> v.matchType()
```

blackwinter commented on metafix/src/main/java/org/metafacture/metafix/FixPath.java:
```diff
@@ -0,0 +1,381 @@
+        }
+    }
+
+    /*package-private*/ void transformIn(final Hash hash, final UnaryOperator<String> operator) {
```
Must be `public`. Currently needed in Limetrans. Or what is the intended replacement for `Record.transformFields(List<String>, UnaryOperator<String>)`?

blackwinter reviewed (changes requested):
Overall, nice to see `Value` shrink a little in size :)

fsteeg commented:
Thanks for the review! Addressed your feedback in https://github.com/metafacture/metafacture-fix/pull/133/commits/a9aac45236e56c10d67d3b584fd52df15f95c814.

blackwinter reviewed (approved):
Great, thanks!
